spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-configuration-watcher-it/src/test/java/org/springframework/cloud/kubernetes/configuration/watcher/ActuatorRefreshKafkaIT.java

```diff
@@ -140,6 +141,8 @@ void after() throws Exception {
 		k8SUtils.waitForDeploymentToBeDeleted(SPRING_CLOUD_K8S_CONFIG_WATCHER_IT_DEPLOYMENT_NAME, NAMESPACE);
 	}
 
+	// TODO figure out why this one fails on bus-starter-4.0.0-SNAPSHOT
+	@Disabled
```

Comment:
I am disabling it for now. I've spent a few good days trying to figure out what is going on here, but my limited understanding of bus/binders/functions/kafka/etc did not help. I'll try to spent some more time in the future, as I don't like disabling this at all :(

Reply:
So you can reproduce this locally?  What do I need to do to reproduce it?

Reply:
It's going to be a bit of mouthful to explain how to reproduce this locally, but in the meanwhile I am on to something (I think). May be you know this off the top of your head.

So in `2.1.x`we get a Bean from `BusRefreshAutoConfiguration` called : `RefreshListener`:


```
    @Bean
    @ConditionalOnProperty(
        value = {"spring.cloud.bus.refresh.enabled"},
        matchIfMissing = true
    )
    @ConditionalOnBean({ContextRefresher.class})
    public RefreshListener refreshListener(ContextRefresher contextRefresher, ServiceMatcher serviceMatcher) {
        return new RefreshListener(contextRefresher, serviceMatcher);
    }
```

After this one is invoked, the message ends up in the kafka queue.

What is interested is that this Bean is present in `main` - I can get it from the context, but it's not present in the "applicationListeners" of the app. It's almost like there should be some code that registers it, but it's not. I'm sorry if this is obscure explanation, but I am still trying to figure out...

Reply:
Weird, the autoconfiguration runs in main though?

Reply:
right, it does:

```
2022-08-09T18:25:56.563+03:00 TRACE 14798 --- [           main] o.s.b.a.condition.OnPropertyCondition    : Condition OnPropertyCondition on org.springframework.cloud.bus.BusRefreshAutoConfiguration#refreshListener matched due to @ConditionalOnProperty (spring.cloud.bus.refresh.enabled) matched
```

when I start the app from `main`

Reply:
Exactly. It is present as a Bean, but never called. This has to be either a race somewhere or a recent change in boot? snapshots, because this test used to work, in this PR.

Reply:
OK if you can recreate something that reproduces the problem fairly easily I can track down the right people to figure it out.

Reply:
Let me show you how I reproduce it and let's see if it works.

- start zookeeper locally: 

```
docker run --rm -it -p 2181:2181 --name=zookeeper -e ZOOKEEPER_CLIENT_PORT=2181 -e ZOOKEEPER_ID=1 -e ZOOKEEPER_SERVER_1=zookeeper zookeeper:3.6.2
```

- find it's IP:

```
docker inspect zookeeper --format='{{ .NetworkSettings.IPAddress }}'
```

- start kafka locally:

```
docker run --rm -it -p 9092:9092 -e KAFKA_ADVERTISED_PORT=9092 -e KAFKA_ADVERTISED_HOST_NAME=localhost -e KAFKA_ZOOKEEPER_CONNECT='172.17.0.2:2181'  wurstmeister/kafka:2.13-2.6.0
```

`172.17.0.2` comes from the previous command.

- start a local kind cluster:

```
kind create cluster
```

- build `spring-cloud-kubernetes-commons` project:

```
cd spring-cloud-kubernetes-commons/
mvn clean install -Dskip.build.image=true -DskipTests -U
```

- build `spring-cloud-kubernetes-client-config`:

```
cd spring-cloud-kubernetes-client-config
mvn clean install -Dskip.build.image=true -DskipTests -U
```

Everything is ready so that our app can start.  Run `ConfigMapWatcherChangeDetector` (from `spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher`) with 3 env variables:

```
SPRING_MAIN_CLOUDPLATFORM=kubernetes;SPRING_PROFILES_ACTIVE=bus-kafka;SPRING_CLOUD_KUBERNETES_CLIENT_NAMESPACE=default
```

App is now ready to react to changes, so let's create a configmap: 

```
apiVersion: v1
data:
    key: value
kind: ConfigMap
metadata:
  name: spring-cloud-kubernetes-configuration-watcher
  labels:
    spring.cloud.kubernetes.config: "true"
```

This in turn, should trigger a kafka message in `springCloudBus` topic. To see if it's really there:

```
docker ps
// get the hash of the kafka image
docker exec -it <PREVIOUS_HASH> bash
/op/kafka/bin/kafka-console-consumer.sh --topic springCloudBus --from-beginning --bootstrap-server localhost:9092
```

Reply:
do you think its safe to merge this the way it is and to track the test failure in an individual issue? this PR was supposed to be much simpler and grew to things I never expected. :)

Reply:
Im just looking now and noticing our jenkins builds are failing as well.  We are going to have to look into this some more.

Reply:
if you find out what is going on, please ping me also. I am dying to find out

Reply:
I'm confused by your example here to reproduce the problem.  The name of the configmap is `spring-cloud-kubernetes-configuration-watcher` wouldn't that then target a refresh of the config watcher itself?  Shoudn't we be targeting a refresh of another app?

What I see happening in your example is that because the rerfesh is for the config watcher itself it shuts down the configwatcher context.  That is not what is happening in the integration test.

Reply:
My bad! Its an incorrect name, it must have the `-it` at the end: `spring-cloud-kubernetes-configuration-watcher-it`. Sorry, I had too many changes in motion.

Reply:
I edited the comment. Sorry again.

Reply:
Btw, it does not even matter if that name is wrong. You would still not see a message in the kafka queue, and that is the main problem. If you look at my example, I do not even have a consumer ( there is no `spring-cloud-kubernetes-configuration-watcher-it` deployed ).

Reply:
I also do not know if you noticed, but the problem is for sure in a SNAPSHOT version. While this test was failing, the circleci one was passing - and we had a bug there where the cache was not updating with the latest snapshots.

Reply:
I was talking to some people yesterday and there were some changes in Kafka recently that seems to correlate with the timeframe in which we saw these failures.  I am looking into it some more.  Hopefully will have an update next week.

Reply:
Thank you. Its actually good that we started this task, we found out about at least 3 issues, which I will fix.

Reply:
OK so I finally nailed down that was happening after I was able to reproduce the problem using a simple Spring Cloud Bus + Kafka app.  This exception was being thrown when we called `triggerRefresh`

```
java.lang.NoSuchMethodError: 'org.springframework.util.concurrent.ListenableFuture org.springframework.kafka.core.KafkaTemplate.send(org.apache.kafka.clients.producer.ProducerRecord)'
	at org.springframework.integration.kafka.outbound.KafkaProducerMessageHandler.handleRequestMessage(KafkaProducerMessageHandler.java:513) ~[spring-integration-kafka-6.0.0-M4.jar:6.0.0-M4]
```
The reason why we didn't see it is because the ExecutorService was swallowing it 🤦‍♂️.  

I added a try/catch with a log statement around that call so we can catch anything like this in the future.

https://github.com/spring-cloud/spring-cloud-kubernetes/blob/17b61c22d32d51b7c554060f179a1447630414a4/spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/ConfigMapWatcherChangeDetector.java#L69

The reason for the NoSuchMethodError was because there was a chance in Spring Core to `ListenableFuture` but the Spring Boot snapshots were using Spring Integration Milestone 4 which didn't have the corresponding change to support the change to `ListenableFuture`.  I got the Boot team to move Spring Integration to snapshots and that fixed the underlying problem.

So we should be able to remove the `@Ignore` from the test now and that build should pass.

Reply:
NICE!! Indeed it builds just fine now :) take a look when you have time